### PR TITLE
Querybuilder: Add when/unless/tap support

### PR DESCRIPTION
**Pull request by ryanmitchell**
Allows you to write queries such as:

```
        $query = Entry::query()
            ->when(request('something', 0) == 1, function($query) {
                $query->where('slug', 'my-slug');
            })
            ->get();

        $query = Entry::query()
            ->unless(request('something', 0) == 1, function($query) {
                $query->where('slug', 'my-slug');
            })
            ->get();
```

which makes code a lot cleaner, especially in query scopes.

The code is directly lifted from Laravel (https://github.com/laravel/framework/blob/fbf25e13b7ce9847fcc336bb3125933638acb936/src/Illuminate/Database/Concerns/BuildsQueries.php#L277) and I brought `tap()` over too as I'm sure someone might have a use for it

### CODE REVIEW

File: src/Query/Builder.php [hunk 1]
```diff
@@ -304,6 +304,33 @@ abstract public function count();
 
     abstract public function get($columns = ['*']);
 
+    public function when($value, $callback, $default = null)
+    {
+        if ($value) {
+            return $callback($this, $value) ?: $this;
+        } elseif ($default) {
+            return $default($this, $value) ?: $this;
```
Comment (edalzell) [inline src/Query/Builder.php:312]: Can we do an early return here instead of the `elseif`?

File: src/Query/Builder.php [hunk 2]
```diff
@@ -304,6 +304,33 @@ abstract public function count();
...
+    public function tap($callback)
+    {
+        return $this->when(true, $callback);
+    }
+
+    public function unless($value, $callback, $default = null)
+    {
+        if (! $value) {
+            return $callback($this, $value) ?: $this;
+        } elseif ($default) {
```
Comment (edalzell) [inline src/Query/Builder.php:327]: same here, early return?

Comment (ryanmitchell) [reply src/Query/Builder.php:327]: I don't think they can as that affects the sequence of logic. As I said its been lifted straight from the laravel codebase.

Comment (edalzell) [reply src/Query/Builder.php:327]: how can it affect the logic if there's a `return`?

Comment (ryanmitchell) [reply src/Query/Builder.php]: I've updated it now - ran the tests and seems ok.

**Comment by jasonvarga** [comment]
I tried using the `Conditionable` trait. Worked nicely, except it didn't exist on older versions of Laravel. Bummer. Maybe you already tried that. 😄 

Thanks for the PR!

**Pull request merged**